modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/internal/upgrade/v4_1_0/KBArticleExternalReferenceCodeUpgradeProcess.java

```diff
@@ -0,0 +1,67 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.knowledge.base.internal.upgrade.v4_1_0;
+
+import com.liferay.knowledge.base.internal.upgrade.v4_1_0.util.KBArticleTable;
+import com.liferay.portal.kernel.dao.jdbc.AutoBatchPreparedStatementUtil;
+import com.liferay.portal.kernel.upgrade.UpgradeProcess;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+
+/**
+ * @author Luis Miguel Barcos
+ */
+public class KBArticleExternalReferenceCodeUpgradeProcess
+	extends UpgradeProcess {
+
+	@Override
+	protected void doUpgrade() throws Exception {
+		if (!hasColumn(KBArticleTable.TABLE_NAME, "externalReferenceCode")) {
+			alter(
+				KBArticleTable.class,
+				new AlterTableAddColumn(
+					"externalReferenceCode", "VARCHAR(75)"));
+		}
+
+		_populateExternalReferenceCode();
+	}
+
+	private void _populateExternalReferenceCode() throws Exception {
+		try (PreparedStatement preparedStatement1 = connection.prepareStatement(
```

Comment:
Can we do it in one query?

Like `update KBArticle set externalReferenceCode = ? where kbArticleId = ? and externalReferenceCode is null or externalReferenceCode = ''`

@javierdearcos

Reply:
I thought about it and I didn't figure out how to do it.

I've ended with a query much like your query, but I've got stuck. Because in the query that we are using we are replacing the value of kbArticleId of the second query iterating from the resultSet of the first query where you have the kbArticleIds, but I don't know how we can make the same with one query

Reply:
Why not this one?
`update KBArticle set externalReferenceCode = CAST_TEXT(kbArticleId) where externalReferenceCode is null or externalReferenceCode = ''`

Reply:
That one makes total sense, yes. I was dumb thinking only in placeholders...

Reply:
We just have to be careful with the syntax in `CAST` / `CONVERT` / `CAST_TEXT` that has to be supported in all major databases. I've seen `CAST_TEXT` in other upgrades so it seems safe but double check looking at other upgrade files.